lib/Frontend/DiagnosticHelper.cpp

```diff
@@ -0,0 +1,522 @@
+
+    // If we got this far, we need to suppress the output of the
+    // PrintingDiagnosticConsumer to ensure that only the parseable-output
+    // is emitted
+    PDC.setSuppressOutput(true);
+  }
+
+  // Because the serialized diagnostics consumer is initialized here,
+  // diagnostics emitted above, within CompilerInvocation::parseArgs, are never
+  // serialized. This is a non-issue because, in nearly all cases, frontend
+  // arguments are generated by the driver, not directly by a user. The driver
+  // is responsible for emitting diagnostics for its own errors.
+  // See https://github.com/apple/swift/issues/45288 for details.
+  SerializedConsumerDispatcher = createSerializedDiagnosticConsumerIfNeeded(
+      invocation.getFrontendOptions().InputsAndOutputs,
+      invocation.getDiagnosticOptions().EmitMacroExpansionFiles);
+  if (SerializedConsumerDispatcher)
+    instance.addDiagnosticConsumer(SerializedConsumerDispatcher.get());
+
+  FixItsConsumer = createJSONFixItDiagnosticConsumerIfNeeded(invocation);
+  if (FixItsConsumer)
+    instance.addDiagnosticConsumer(FixItsConsumer.get());
+
+  if (invocation.getDiagnosticOptions().UseColor)
+    PDC.forceColors();
+
+  PDC.setPrintEducationalNotes(
+      invocation.getDiagnosticOptions().PrintEducationalNotes);
+
+  PDC.setFormattingStyle(
+      invocation.getDiagnosticOptions().PrintedFormattingStyle);
+
+  PDC.setEmitMacroExpansionFiles(
+      invocation.getDiagnosticOptions().EmitMacroExpansionFiles);
+}
+
+static const char *
+mapFrontendInvocationToAction(const CompilerInvocation &Invocation) {
+  FrontendOptions::ActionType ActionType =
+      Invocation.getFrontendOptions().RequestedAction;
+  switch (ActionType) {
+  case FrontendOptions::ActionType::REPL:
+    return "repl";
+  case FrontendOptions::ActionType::MergeModules:
+    return "merge-module";
+  case FrontendOptions::ActionType::Immediate:
+    return "interpret";
+  case FrontendOptions::ActionType::TypecheckModuleFromInterface:
+    return "verify-module-interface";
+  case FrontendOptions::ActionType::EmitPCH:
+    return "generate-pch";
+  case FrontendOptions::ActionType::EmitIR:
+  case FrontendOptions::ActionType::EmitBC:
+  case FrontendOptions::ActionType::EmitAssembly:
+  case FrontendOptions::ActionType::EmitObject:
+    // Whether or not these actions correspond to a "compile" job or a
+    // "backend" job, depends on the input kind.
+    if (Invocation.getFrontendOptions().InputsAndOutputs.shouldTreatAsLLVM())
+      return "backend";
+    else
+      return "compile";
+  case FrontendOptions::ActionType::EmitModuleOnly:
+    return "emit-module";
+  default:
+    return "compile";
+  }
+  // The following Driver/Parseable-output actions do not correspond to
+  // possible Frontend invocations:
+  // ModuleWrapJob, AutolinkExtractJob, GenerateDSYMJob, VerifyDebugInfoJob,
+  // StaticLinkJob, DynamicLinkJob
+}
+
+// TODO: Apply elsewhere in the compiler
```

Comment:
👀

Reply:
This is directly transported from original code. Do I own the TODO now that `git blame` will trace to me?